### PR TITLE
VectorizedArrayIterator enable std::advance()

### PR DESCRIPTION
**Pull request by peterrum**
As requested in https://github.com/dealii/dealii/pull/9674#pullrequestreview-376505954.

### CODE REVIEW

File: include/deal.II/base/vectorization.h [hunk 1]
```diff
@@ -169,6 +169,29 @@ class VectorizedArrayIterator
...
   }
 
+  /**
+   * This operator advances the iterator by @p offet lanes and returns a
+   * reference to <tt>*this</tt>.
+   */
+  VectorizedArrayIterator<T> &
+  operator+=(unsigned int offset)
+  {
+    lane += offset;
```
Comment (masterleinad) [inline include/deal.II/base/vectorization.h]: What do you do if `lane` is larger than `T::n_array_elements`?

Comment (peterrum) [reply include/deal.II/base/vectorization.h]: @masterleinad I don't see a problem that `lane` is outside of the range `[0;T::size()[`. In contrast, we need at least `static_cast<unsignd int>(-1)` () and `T::size()` (in the case of incrementing) as `end()`. The problem is only when we want to dereferene the iterator. This is why I have added an assert to `operator*()`. Do you agree that this enough?

Comment (masterleinad) [reply include/deal.II/base/vectorization.h]: Yes, I meant to say that we should make sure that the element referenced stays valid. I think checking in the accessor is fine.

Comment (bangerth) [reply include/deal.II/base/vectorization.h]: I'm usually of the opinion that we should do the checking at the first possible place because it makes debugging so much simpler. So I would advocate putting an assertion here as well.

File: include/deal.II/base/vectorization.h [hunk 2]
```diff
@@ -169,6 +171,29 @@ class VectorizedArrayIterator
     return *this;
   }
 
+  /**
+   * This operator advances the iterator by @p offet lanes and returns a
+   * reference to <tt>*this</tt>.
+   */
+  VectorizedArrayIterator<T> &
+  operator+=(unsigned int offset)
```
Comment (bangerth) [inline include/deal.II/base/vectorization.h]:
```suggestion
  operator+=(const unsigned int offset)
```

File: include/deal.II/base/vectorization.h [hunk 3]
```diff
@@ -169,6 +171,29 @@ class VectorizedArrayIterator
...
+  /**
+   * Prefix <tt>--</tt> operator: <tt>--iterator</tt>. This operator advances
+   * the iterator to the previous lane and returns a reference to
+   * <tt>*this</tt>.
+   */
+  VectorizedArrayIterator<T> &
+  operator--()
+  {
+    lane--;
+    return *this;
```
Comment (bangerth) [inline include/deal.II/base/vectorization.h]: That would suggest also putting an assertion here. (And possibly in `operator++`, which is not part of this patch.)

**Comment by peterrum** [comment]
@masterleinad @bangerth Thanks for the review. I have made the changes.

File: include/deal.II/base/vectorization.h [hunk 4]
```diff
@@ -165,16 +167,41 @@ class VectorizedArrayIterator
   VectorizedArrayIterator<T> &
   operator++()
   {
-    lane++;
+    AssertIndexRange(lane + 1, T::size() + 1) lane++;
```
Comment (bangerth) [inline include/deal.II/base/vectorization.h]: Not wrong but hard to read :-)
```suggestion
    AssertIndexRange(lane + 1, T::size() + 1);
    ++lane;
```

File: include/deal.II/base/vectorization.h [hunk 5]
```diff
@@ -165,16 +167,41 @@ class VectorizedArrayIterator
...
+  }
+
+  /**
+   * This operator advances the iterator by @p offet lanes and returns a
+   * reference to <tt>*this</tt>.
+   */
+  VectorizedArrayIterator<T> &
+  operator+=(const unsigned int offset)
+  {
+    AssertIndexRange(lane + offset, T::size() + 1) lane += offset;
```
Comment (bangerth) [inline include/deal.II/base/vectorization.h]:
```suggestion
    AssertIndexRange(lane + offset, T::size() + 1);
    lane += offset;
```

File: include/deal.II/base/vectorization.h [hunk 6]
```diff
@@ -165,16 +167,41 @@ class VectorizedArrayIterator
...
+  /**
+   * Prefix <tt>--</tt> operator: <tt>--iterator</tt>. This operator advances
+   * the iterator to the previous lane and returns a reference to
+   * <tt>*this</tt>.
+   */
+  VectorizedArrayIterator<T> &
+  operator--()
+  {
+    AssertIndexRange(lane, T::size() + 1);
+    lane--;
```
Comment (bangerth) [inline include/deal.II/base/vectorization.h]:
```suggestion
    --lane;
```

File: include/deal.II/base/vectorization.h [hunk 7]
```diff
@@ -165,16 +167,41 @@ class VectorizedArrayIterator
...
+
+  /**
+   * Prefix <tt>--</tt> operator: <tt>--iterator</tt>. This operator advances
+   * the iterator to the previous lane and returns a reference to
+   * <tt>*this</tt>.
+   */
+  VectorizedArrayIterator<T> &
+  operator--()
+  {
+    AssertIndexRange(lane, T::size() + 1);
```
Comment (bangerth) [inline include/deal.II/base/vectorization.h]: This is wrong. It should have been
```suggestion
    Assert(lane>0, ExcMessage ("You can't decrement an iterator that is already at the beginning of the range.");
```

**Comment by peterrum** [comment]
@bangerth Thanks! Done!
